### PR TITLE
Fixes #127

### LINKED ISSUE #127: Add function to calculate timing and size of epidemic peak
This issue comes from comments on #33 (https://github.com/epiverse-trace/epidemics/issues/33#issuecomment-1751216372), and is to request the addition of functions that find and calculate the size and timing of a peak in any of the compartments in an epidemic.

The functionality should include the option to calculate the peak separately for demographic groups. This function may require the use of change-point detection algorithms, or multivariate segmentation clustering algorithms.

A vignette should be added for this functionality and should include examples where the peak is not easy to define, such as when implementing interventions that reduce transmission ("flattening the curve").

### PR DESCRIPTION
**Pull request by bahadzie**
- Fixes issue #125
- Fixes issue #127

### CODE REVIEW

**Comment by Bisaloo** [comment]
Hi @bahadzie, rather than removing the files in `_snaps/` altogether, you should run the tests interactively on your local machine, and run `testthat::snapshot_accept('vaccination')` if you want to validate the change in the recorded snapshot.

More info in the dedicated testthat vignette: https://testthat.r-lib.org/articles/snapshotting.html

**Comment by pratikunterwegs** [comment]
Hi @bahadzie and @Bisaloo - just checking whether this PR is still current and if it needs to be reviewed? @Bisaloo are you reviewing or would it help for me to step in?

**Comment by Bisaloo** [comment]
No, this was just a drive by comment based on the check results. I have not reviewed the code. I'll leave it up to your expertise.

**Comment by pratikunterwegs** [comment]
Okay thanks, we have a P3 meeting scheduled later so will discuss next steps for this PR then.

**Review by pratikunterwegs — CHANGES_REQUESTED** [review]
Hi @bahadzie, thanks for making this PR and taking a stab at this issue.

This function works mostly alright for the case of a single demographic group in a model where time increments by 1.0, but would not find the correct epidemic peak time for time increments other than 1.0 (I think), nor for multiple demographic groups. This is because of how `timing <- which(wide %in% peak)` is calculated - you're getting the index, not the time value. I've attached an example from the vignettes below - the figure shows all three peaks at about 370, but the function calculates peaks of 367, 972, and 1576. I've made some suggestions for edits to the function which should make it more readable and perform as expected.

- [ ] This function can be collapsed, for its simplest case in a single-peak epidemic, by using convenient {data.table} syntax from the suggestion.
- [ ] You should also add tests for this function. There should be tests for the output type, dimensions, and predictable column contents (e.g. demography group names) and types (timing and value should be numerics).
- [ ] You should revert commit https://github.com/epiverse-trace/epidemics/pull/152/commits/9a349b2eb21dd0bf59c4fc6da4eabc79f985e6fd, as these snapshots are important and must be retained.

``` r
library(epidemics)
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(ggplot2)

# load contact and population data from socialmixr::polymod
polymod <- socialmixr::polymod
contact_data <- socialmixr::contact_matrix(
  polymod,
  countries = "United Kingdom",
  age.limits = c(0, 20, 40),
  symmetric = TRUE
)
#> Using POLYMOD social contact data. To cite this in a publication, use the 'cite' function
#> Removing participants that have contacts without age information. To change this behaviour, set the 'missing.contact.age' option

# prepare contact matrix
contact_matrix <- t(contact_data$matrix)

# prepare the demography vector
demography_vector <- contact_data$demography$population
names(demography_vector) <- rownames(contact_matrix)

# view contact matrix and demography
contact_matrix
#>                  
#> contact.age.group     [,1]     [,2]     [,3]
#>           [0,20)  7.883663 2.794154 1.565665
#>           [20,40) 3.120220 4.854839 2.624868
#>           40+     3.063895 4.599893 5.005571

demography_vector
#>   [0,20)  [20,40)      40+ 
#> 14799290 16526302 28961159
# initial conditions
initial_i <- 1e-6
initial_conditions <- c(
  S = 1 - initial_i, E = 0, I = initial_i, R = 0, V = 0
)

# build for all age groups
initial_conditions <- rbind(
  initial_conditions,
  initial_conditions,
  initial_conditions
)

# assign rownames for clarity
rownames(initial_conditions) <- rownames(contact_matrix)

# view initial conditions
initial_conditions
#>                S E     I R V
#> [0,20)  0.999999 0 1e-06 0 0
#> [20,40) 0.999999 0 1e-06 0 0
#> 40+     0.999999 0 1e-06 0 0

uk_population <- population(
  name = "UK",
  contact_matrix = contact_matrix,
  demography_vector = demography_vector,
  initial_conditions = initial_conditions
)

uk_population
#> <population> object
#> 
#>  Population name:
#> "UK"
#> 
#>  Demography 
#> [0,20): 14,799,290 (20%)
#> [20,40): 16,526,302 (30%)
#> 40+: 28,961,159 (50%)
#> 
#>  Contact matrix 
#>                  
#> contact.age.group   [0,20)  [20,40)      40+
#>           [0,20)  7.883663 2.794154 1.565665
#>           [20,40) 3.120220 4.854839 2.624868
#>           40+     3.063895 4.599893 5.005571

# run epidemic simulation with no vaccination or intervention
data <- model_default_cpp(
  population = uk_population,
  time_end = 600, increment = 1.0
)

# plot figure of epidemic curve
filter(data, compartment %in% c("exposed", "infectious")) %>%
  ggplot(
    aes(
      x = time,
      y = value,
      col = demography_group,
      linetype = compartment
    )
  ) +
  geom_line() +
  labs(
    x = "Simulation time (days)",
    linetype = "Compartment",
    y = "Individuals"
  )
```



``` r

# get the timing and peak of the infectious compartment
epidemic_peak(data)
#>         timing     peak
#> [0,20)     367 384139.9
#> [20,40)    972 357215.9
#> 40+       1576 474553.3
```

<sup>Created on 2024-01-22 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

File: R/helpers.R [hunk 1]
```diff
@@ -298,3 +298,65 @@ new_infections <- function(data,
...
+#' data <- model_default_cpp(
+#'   population = uk_population
+#' )
+#'
+#' # get the timing and peak of the infectious compartment
+#' epidemic_peak(data)
+#'
+#' # get the timing and peak of the exposed compartment
+#' epidemic_peak(data, compartment = "exposed")
+epidemic_peak <- function(data, compartment = "infectious") {
```
[the remark this answers is not in the document]
Comment (bahadzie) [reply R/helpers.R:335]: Done

File: R/helpers.R [hunk 2]
```diff
@@ -298,3 +298,65 @@ new_infections <- function(data,
...
+#'   demography_vector = 67e6,
+#'   initial_conditions = matrix(
+#'     c(0.9999, 0.0001, 0, 0, 0),
+#'     nrow = 1, ncol = 5L
+#'   )
+#' )
+#'
+#' # run epidemic simulation with no vaccination or intervention
+#' data <- model_default_cpp(
+#'   population = uk_population
```
[the remark this answers is not in the document]
Comment (bahadzie) [reply R/helpers.R:327]: Done

File: R/helpers.R [hunk 3]
```diff
@@ -298,3 +298,65 @@ new_infections <- function(data,
...
+  # check parameters are as expected
+  checkmate::assert_data_frame(
+    data,
+    min.cols = 4L, min.rows = 1, any.missing = FALSE
+  )
+  checkmate::assert_names(
+    colnames(data),
+    must.include = c("time", "demography_group", "compartment", "value")
+  )
+  checkmate::assert_character(compartment, len = 1L)
```
[the remark this answers is not in the document]
Comment (bahadzie) [reply R/helpers.R:345]: Done

File: R/helpers.R [hunk 4]
```diff
@@ -298,3 +298,65 @@ new_infections <- function(data,
...
+  nrows <- length(long_data[, "value"])
+  #  long to wide conversion
+  wide <- matrix(long_data[, "value"], nrows / n_groups, n_groups, byrow = TRUE)
+  colnames(wide) <- age_groups
+
+  col_max <- function(data) sapply(as.data.frame(data), max, na.rm = TRUE)
+  peak <- col_max(as.data.frame(wide))
+  timing <- which(wide %in% peak)
+
+  cbind(timing, peak)
```
[the remark this answers is not in the document]
Comment (bahadzie) [reply R/helpers.R:361]: Done. Although I get a note about undefined global functions or variables `compartment`, `time` and `value`.

Not sure how to adjust the `data.table` syntax to pass `lintr` rules

**Review by pratikunterwegs — CHANGES_REQUESTED** [review]
Hi @bahadzie, thanks for adding these changes. Could you:
1. Update the PR description to say what the changes are and ideally tag the issues being fixed;
2. Rebase this branch on `main`;
3. Update your version of {testthat} (v3.2.1 appears latest) as this seems to be causing changes to the test snapshots; you would also need to revert changes to the test snapshots, and then check if any snapshots really need to be updated. In general, I would be very conservative in accepting changes to a test snapshot, and I would always mention why in the PR or in the commit message;
4. Update the code in `epidemic_peak()` to access columns using [one of the two methods mentioned in this SO post](https://stackoverflow.com/a/8096882/7410366) by Matt Dowle. In today's meeting I said I preferred setting variables to `NULL` at the top of the function (Dowle's option 2), but I see that in `new_infections()` I have gone for option 1 instead which uses `get()`.

Once 2 -- 4 are done, we shall get a better idea of whether checks pass - I expect they will.

**Comment by bahadzie** [comment]
> Hi @bahadzie, thanks for adding these changes. Could you:
> 
>     1. Update the PR description to say what the changes are and ideally tag the issues being fixed;
> 
>     2. Rebase this branch on `main`;
> 
>     3. Update your version of {testthat} (v3.2.1 appears latest) as this seems to be causing changes to the test snapshots; you would also need to revert changes to the test snapshots, and then check if any snapshots really need to be updated. In general, I would be very conservative in accepting changes to a test snapshot, and I would always mention why in the PR or in the commit message;
> 
>     4. Update the code in `epidemic_peak()` to access columns using [one of the two methods mentioned in this SO post](https://stackoverflow.com/a/8096882/7410366) by Matt Dowle. In today's meeting I said I preferred setting variables to `NULL` at the top of the function (Dowle's option 2), but I see that in `new_infections()` I have gone for option 1 instead which uses `get()`.
> 
> 
> Once 2 -- 4 are done, we shall get a better idea of whether checks pass - I expect they will.

**Comment by bahadzie** [comment]
> Hi @bahadzie, thanks for adding these changes. Could you:
> 
>     1. Update the PR description to say what the changes are and ideally tag the issues being fixed;
> 
>     2. Rebase this branch on `main`;
> 
>     3. Update your version of {testthat} (v3.2.1 appears latest) as this seems to be causing changes to the test snapshots; you would also need to revert changes to the test snapshots, and then check if any snapshots really need to be updated. In general, I would be very conservative in accepting changes to a test snapshot, and I would always mention why in the PR or in the commit message;
> 
>     4. Update the code in `epidemic_peak()` to access columns using [one of the two methods mentioned in this SO post](https://stackoverflow.com/a/8096882/7410366) by Matt Dowle. In today's meeting I said I preferred setting variables to `NULL` at the top of the function (Dowle's option 2), but I see that in `new_infections()` I have gone for option 1 instead which uses `get()`.
> 
> 
> Once 2 -- 4 are done, we shall get a better idea of whether checks pass - I expect they will.

Done

**Comment by pratikunterwegs** [comment]
Thanks, will take a look tomorrow, checks are passing.

**Review by pratikunterwegs — CHANGES_REQUESTED** [review]
Thanks for these inputs @bahadzie. Just looking at @BlackEdder's comments on #125, I think the expectation is for a function like `expected_vaccinations(population, vaccination)`, which calculates the expected number of vaccinated individuals in each age group, before/without running an epidemic model. This would only work for a single dose model, so not Vacamole - although perhaps I'm wrong here, and this could be done as $\nu_1 \nu_2 * S$?

The calculation of an effective vaccination rate should be an internal step, as it determines the transition rates within the model, so it's not expected to an exported function as far as I understand.

@BlackEdder, could you come in and say whether it's necessary to calculate an effective vaccination rate internally? Currently, users pass cohort-specific, per-interval (1 day) vaccination rates ($\nu_i$), which is applied only to susceptibles: $V_i = \nu_i S_i$. Or have I not understood this issue correctly?

**Comment by pratikunterwegs** [comment]
Just looking at this PR again, I think it might be better to restrict this PR to the epidemic peak function by resetting this PR to before the vaccination rate fn is added, I expect that should be https://github.com/epiverse-trace/epidemics/pull/152/commits/8bee33369235d994219efdddce87d3ac289adb50. I think #202 should resolve the effective vaccination issue documented in #198.

**Comment by pratikunterwegs** [comment]
Key commits from this branch copied into #240 and merged into `main`.

**Pull request closed without merging**